common/map_test.cpp

```diff
@@ -226,6 +226,92 @@ TYPED_TEST(MapTest, Conversions) {
   EXPECT_EQ(104, *cmv3[4]);
 }
 
+TYPED_TEST(MapTest, Grow) {
+  using MapT = TypeParam;
+
+  MapT m;
+  // Grow when empty. May be a no-op for some small sizes.
+  m.Grow(32);
+
+  // Add some elements that will need to be propagated through subsequent
+  // growths. Also delete some.
+  for (int i : llvm::seq(1, 24)) {
+    SCOPED_TRACE(llvm::formatv("Key: {0}", i).str());
+    ASSERT_TRUE(m.Insert(i, i * 100).is_inserted());
+  }
+  for (int i : llvm::seq(1, 8)) {
+    SCOPED_TRACE(llvm::formatv("Key: {0}", i).str());
+    ASSERT_TRUE(m.Erase(i));
+  }
+
+  // No-op.
+  m.Grow(16);
+  ExpectMapElementsAre(
+      m, MakeKeyValues([](int k) { return k * 100; }, llvm::seq(8, 24)));
+
+  // Get a couple of doubling based growths.
+  m.Grow(64);
```

Comment:
`CountProbedKeys` would go down here, in a deterministic way depending how you manipulated the hash function.

Reply:
See above about metrics for testing that probing improves.

I can switch to the forced-collision hash if you think its important to test probing reduction here, but it seemed better to get coverage from the variety of different key/value types and small sizes.

Reply:
The metrics based assertions are done. Not clear if you think its important to also have the forced-collision side of this, let me know.

Reply:
The metrics approach is fine. I suggested in another comment a test to show that tombstones were cleared out by growth. It would be nice to show `CountProbedKeys` going down when growing (since that is one of the benefits of growth and it not going down could be indicative of a bug that affects performance but not correctness), but not necessary.

Reply:
I feel like for both the probed keys and the tombstones, this isn't really about testing the behavior of these APIs, but of the growth routines more generally... And that's proven a bit tricky for me to do.

I tried adding the probed key count test, but we end up with zero probed keys here in some cases, and so the count doesn't change. And I left a comment in the other thread about why I'm not really sure how to test the tombstone removal without ending up over-fitting to the particular implementation.

I also tried adding a test for probed key counts decreasing with the forced hash collisions, and that actually doesn't work either -- because we force the hashes to collide, even growth doesn't remove the collision! =[

The only way I see to test this behavior realistically is to contrive a synthetic hash that has collisions below some number of bits but not above, and then notice that once we grow past that threshold, the probing decreases. But this seems much too brittle to me, and the benchmarks seem like a better place to make sure we don't have a performance problem here.

So I've tried a bunch of things here that haven't worked out. Let me know if you still see something that seems important to add coverage for.